pybossa/model/task.py

```diff
@@ -75,4 +75,26 @@ def pct_status(self):
         Index('task_info_idx', sqlalchemy.text('md5(info::text)')),
     )
 
+    @staticmethod
+    def apply_access_control(data, user=None, project_data=None):
+        def has_full_access():
+            if not user:
+                return False
+            if user.admin:
+                return True
+            if not user.subadmin:
+                return False
+            if user.id not in project_data['owners_ids']:
+                return False
+            return True
```

Comment:
Just a thought. What are the cases where a user should have access to this data?

Since this involves allowing access to gold answer data, by default, perhaps the result should be False. Only explicitly defined types of users would be permitted access (i.e., `user.admin`).

In the above case, the code could be summarized as:

```python
return user.admin
```

Reply:
See if decorator only_admin_or_subadminowner can be used here that is defined [here](https://github.com/bloomberg/pybossa/blob/master/pybossa/auth/project.py#L32)

Reply:
@primaryobjects, the cases where the user has access to the data are spelled out in the function `has_full_access()`. How would you write this?

Reply:
The only explicit case where the user has full access appears to be when `user.admin = True`. There is also a default of returning `True`, which maybe we should consider defaulting to `False` to err on the side of data safety? If this is the case, the method could be optimized to just `return user.admin` (if only admins should have full access).

If other users should also have full access, then perhaps we could define those?

Reply:
@primaryobjects, please look at the function and notice that it is doing more than just checking `user.admin`. You comment does not correspond to the code I have written.

Reply:
It looks good as it is. My thought was just to consider restructuring the function to check for `True` cases, rather than mixed (example below), but perhaps the original is best. :)

```python
return user and (user.admin or user.subadmin or user.id in project_data['owners_ids'])
```

Reply:
seems what we mainly need is list of owner ids under func/decorator admin_or_subadmin_owner. If its not too much work, I think you may be able to update it in proj auth passing list instead of proj object and reuse here.

Reply:
I considered doing something like that, but I rejected it in favor of the code I wrote. The reason is that I have found that complex nested boolean expressions are hard to understand and maintain and can lead to bugs. So I "unrolled" the complex nested boolean expression into a series of simple cases like a flow chart or decision tree. It is a pattern I try to use when I can and although it may appear strange I think it is superior to the alternative.

Reply:
@dchhabda, I'll do that.